pod/perlguts.pod

```diff
@@ -153,37 +153,73 @@ Perl's own functions typically add a trailing C<NUL> for this reason.
 Nevertheless, you should be very careful when you pass a string stored
 in an SV to a C function or system call.
 
-To access the actual value that an SV points to, you can use the macros:
-
-    SvIV(SV*)
-    SvUV(SV*)
-    SvNV(SV*)
-    SvPV(SV*, STRLEN len)
-    SvPV_nolen(SV*)
-
-which will automatically coerce the actual scalar type into an IV, UV, double,
-or string.
-
-In the C<SvPV> macro, the length of the string returned is placed into the
-variable C<len> (this is a macro, so you do I<not> use C<&len>).  If you do
-not care what the length of the data is, use the C<SvPV_nolen> macro.
-Historically the C<SvPV> macro with the global variable C<PL_na> has been
-used in this case.  But that can be quite inefficient because C<PL_na> must
+To access the actual value that an SV points to, Perl's API exposes
+several macros that coerce the actual scalar type into an IV, UV, double,
+or string:
+
+=over
+
+=item * C<SvIV(SV*)> (C<IV>) and C<SvUV(SV*)> (C<UV>)
+
+=item * C<SvNV(SV*)> (C<double>)
+
+=item * Strings are a bit complicated:
+
+=over
+
+=item * Byte string: C<SvPVbyte(SV*, STRLEN len)> or C<SvPVbyte_nolen(SV*)>
+
+If the Perl string is C<"\xff\xff">, then this returns a 2-byte C<char*>.
+
+=item * UTF-8 string: C<SvPVutf8(SV*, STRLEN len)> or C<SvPVutf8_nolen(SV*)>
+
+If the Perl string is C<"\xff\xff">, then this returns a 4-byte C<char*>.
+
+=item * You can also use C<SvPV(SV*, STRLEN len)> or C<SvPV_nolen(SV*)>
+to fetch the SV's raw internal buffer. This is tricky, though; if your Perl
+string
+is C<"\xff\xff">, then depending on the SV's internal encoding you might get
+back a 2-byte B<OR> a 4-byte C<char*>.
+Moreover, if it's the 4-byte string, that could come from either Perl
+C<"\xff\xff"> stored UTF-8 encoded, or Perl C<"\xc3\xbf\xc3\xbf"> stored
+as raw octets. To differentiate between these you B<MUST> look up the
+SV's UTF8 bit (cf. C<SvUTF8>) to know whether the source Perl string
+is 2 characters (C<SvUTF8> would be on) or 4 characters (C<SvUTF8> would be
+off).
+
+B<IMPORTANT:> Use of C<SvPV>, C<SvPV_nolen>, or
```

Comment:
This ticket caused me to go through my CPAN modules to update them. And while some of them did need updating (in particular `Crypt::` ones), many did not. My experience is not that `SvPV` was always buggy for a simple reason: it all doesn't matter if your input is ascii.

And frankly, in a lot of cases allowed input is really just ascii.

Reply:
Perhaps then "is almost certainly a bug if non-ASCII input is allowed." But using either SvPVbyte or SvPVutf8 in this case is also correct, so I think it's still a good idea not to recommend SvPV.

Reply:
I’ll update it, but FWIW I still think SvPVbyte is the “more correct” choice, even if the input is all ASCII, since the conceptual “leak” is still there with SvPV, even if there happens to be no externally observable difference in behaviour.

Reply:
Yeah, "is almost certainly a bug if non-ASCII input is allowed" would be much better here

Reply:
Change from the prior history:
```
diff --git a/pod/perlguts.pod b/pod/perlguts.pod
index a97435d9ef..c201be0a6f 100644
--- a/pod/perlguts.pod
+++ b/pod/perlguts.pod
@@ -189,7 +189,8 @@ off).

 B<IMPORTANT:> Use of C<SvPV>, C<SvPV_nolen>, or
 similarly-named macros I<without> looking up the SV's UTF8 bit is
-almost certainly a bug.
+a likely source of character-encoding bugs, unless the string in question
+is always fully UTF-8 invariant.
```

Reply:
@Leont I made the above change before I saw your last comment in this thread. Does what I did satisfy your concerns, or should I change to `is almost certainly a bug if non-ASCII input is allowed`? I don’t have a strong opinion, though I _slightly_ prefer “UTF-8 invariant” rather than “ASCII” in light of EBCDIC support.

Reply:
Either is fine to me. The former is more correct, but the latter may be easier to understand for people less deep into character encodings.

Reply:
A parenthetical `(or non-EBCDIC)` would be sufficient for correctness in the latter case IMO

Reply:
I just went with `is almost certainly a bug if non-ASCII input is allowed`. `perlebcdic` kinda makes clear that EBCDIC folks need to “translate” ASCII in Perl’s documentation to EBCDIC anyway.